java/vector/src/main/codegen/templates/DenseUnionReader.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import org.apache.arrow.vector.types.Types.MinorType;
+import org.apache.arrow.vector.types.pojo.Field;
+
+<@pp.dropOutputFile />
+<@pp.changeOutputFile name="/org/apache/arrow/vector/complex/impl/DenseUnionReader.java" />
+
+
+<#include "/@includes/license.ftl" />
+
+        package org.apache.arrow.vector.complex.impl;
+
+<#include "/@includes/vv_imports.ftl" />
+/**
+ * Source code generated using FreeMarker template ${.template_name}
+ */
+@SuppressWarnings("unused")
+public class DenseUnionReader extends AbstractFieldReader {
+
+  private BaseReader[] readers = new BaseReader[43];
```

Comment:
This looks like it is propagating the incompatibity with the IPC specification.  Specifically a union should be able to have more then one vector for each minor type.  is this the case?

Reply:
Are you asking this question: UnionVector and DenseUnionVector are sharing the same minor type? If so, the answer is yes for now.
Do you think we should provide separate minor types for them?

Reply:
No I'm asking if there can be two different vectors of the same type in the union (e.g. two different struct vectors).

Reply:
I see. Thanks for the clarification. 

You are right. This implementation does not support more than one vector for each minor type. 
This is aligning with the current (Sparse) UnionVector. 

Should we resolve this problem?

Reply:
IMO densevector shouldn't propagate the same problems the existing UnionVector has with compatibility.

Reply:
Thanks a lot for your feedback. I will resolve it later.

Reply:
@emkornfield I have updated the code to allow arbitrary relative types for the desne union vector. Please take a look.

Brief change list:
1. For StructVector and ListVector, a custom type id is registered when they are added, if the same relative type has never occured. 
2. the type id of the custom relative types start from MinorType.values().length, and the maximum value is 128.
3. some additional data structures are added to track the existing custom relative types, struct vector, and list vectors.